Add availability of Templates via API

m6121 opened the pull request:
This is a pull request that aims at making template management in elabftw available via API. See also #2310

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/elabftw/elabftw?pullRequest=2409) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/elabftw/elabftw?pullRequest=2409) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/elabftw/elabftw?pullRequest=2409) it.</sub>

m6121 commented:
Hi Nico,
I think this is worth a review, now. Not sure why the circleci is failing at the `Code coverage` step as the tests should work the way they are implemented.
Best regards
Max

NicolasCARPi commented:
Hello,

Can you try to describe all of the changes in english as a bullet point list please?

m6121 commented:
**Features:**

* Separated user templates from common team templates:
  * Team templates are stored in a new column `common_template` into the `teams` table
  * User templates are stored as before in the table `experiments_templates`
* Added locking feature of templates analogue to experiments, i.e. templates can be locked to prevent modification
* Added revisions of templates analogue to experiments, i.e. template updated will create a new revision if applicable, revisions can be displayed and restored
* Added template creation via API analogue to experiments, i.e. create a template and update it using the id.

**Some implementation details:**

* `date` was added and `name` was renamed to `title` in order to align Templates a little more with `AbstractEntity`
* `schema58.sql` contains the corresponding database changes, `structure.sql` contains the same, but also the fix in #2499. 
* The common template modification has been removed from a separate tab in the Admin Panel to the first in order to employ the `Team.php` (`MapInterface`), see discussion at https://github.com/elabftw/elabftw/issues/2310#issuecomment-780405457
* Split up `show-templates-view.html` and `show-template-edit.html` into `templates-view.html` and `templates-edit.html` to re-use view when template is locked

Not yet checked if the modifications affect also #2457

m6121 commented:
One question, that I came across when working on this was if this should not point to the "old" body as this is otherwise skipping the initial body and creates a revision of the new body: https://github.com/elabftw/elabftw/blob/726b7ee406f8d13ed17160a151b5f83613ccb685/src/models/AbstractEntity.php#L347-L349

NicolasCARPi commented:
Please try and fix the conflicts. Ask me if you have troubles.

m6121 commented:
Hi Nico,
just solved the conflicts and reached the same point in circle ci, failing at the code coverage.
Best regards
Max

NicolasCARPi commented:
Circleci is not failing at code coverage, it's failing 3 unit tests that seem related to your code (it's green on hypernext branch)

![2021-03-02-232659_686x720_scrot](https://user-images.githubusercontent.com/3043706/109723745-ce43ae00-7bae-11eb-92ca-d74db2170baf.png)


See also: https://scrutinizer-ci.com/g/elabftw/elabftw/inspections/f4dc552a-b0f6-4487-95ae-76d4b426ed76/issues/files/src/models/Templates.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

m6121 commented:
I see.  Unfortunately, I was not able to debug the unit tests locally in order to find the issue since I cannot populate the DB or run the unit tests:
```
bash-5.1# bin/console dev:populate tests/populate-config.yml 
Populate the database with fake data
====================================
Dropping current database and loading structure

In Encoding.php line 66:
                                                   
  Encoding::hexToBin() input is not a hex string.  
                                                   

dev:populate [-u|--smtpuser SMTPUSER] [-s|--smtppass SMTPPASS] [-y|--yes] [--] <file>
```
For testing the environment locally, I created a `phpunit` DB and user and ran `php tests/config-ci.php` before. However, the unit test itself does not seem to find this corresponding configuration.
```
bash-5.1# php vendor/bin/codecept run unit tests/unit/models/TemplatesTest.php 
Codeception PHP Testing Framework v4.1.18
Powered by PHPUnit 9.5.2 by Sebastian Bergmann and contributors.
PHP Warning:  The use statement with non-compound name 'getenv' has no effect in /elabftw/tests/unit/_bootstrap.php on line 9

Unit Tests (7) ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
E TemplatesTest: Create (5.02s)
E TemplatesTest: Read (5.01s)
E TemplatesTest: Get writeable templates list (5.01s)
E TemplatesTest: Duplicate (5.01s)
E TemplatesTest: Read for user (5.01s)
E TemplatesTest: Update (5.00s)
E TemplatesTest: Destroy (5.00s)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 00:35.075, Memory: 10.00 MB

There were 7 errors:

---------
1) TemplatesTest: Create
 Test  tests/unit/models/TemplatesTest.php:testCreate
                                                                                                 
  [PDOException] SQLSTATE[HY000] [2002] php_network_getaddresses: getaddrinfo failed: Try again  
                                                                                                 
#1  /elabftw/src/classes/Db.php:55
#2  /elabftw/src/classes/Db.php:88
#3  /elabftw/src/models/Users.php:45
#4  /elabftw/tests/unit/models/TemplatesTest.php:18

....
```
However, I checked the functionality from the UI which work in both cases so the errors might be related to the implementation of the template tests.

NicolasCARPi commented:
Normally you should be able to run `yarn unit` and it would do everything in new containers. See `tests/run.sh`.

m6121 commented:
Oh, that eases the testing a lot! I just run the `tests/run.sh` script and interestingly, the unit tests run without any error:
```
....
Codeception PHP Testing Framework v4.1.18
Powered by PHPUnit 9.5.2 by Sebastian Bergmann and contributors.
PHP Warning:  The use statement with non-compound name 'getenv' has no effect in /elabftw/tests/unit/_bootstrap.php on line 9

Unit Tests (149) --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
.....
✔ ExperimentsTest: Create and destroy (0.06s)
✔ ExperimentsTest: Set id (0.00s)
✔ ExperimentsTest: Read (0.01s)
✔ ExperimentsTest: Update (0.02s)
✔ ExperimentsTest: Update visibility (0.04s)
✔ ExperimentsTest: Update category (0.02s)
✔ ExperimentsTest: Duplicate (0.03s)
....
✔ TemplatesTest: Create (0.01s)
✔ TemplatesTest: Read (0.00s)
✔ TemplatesTest: Get writeable templates list (0.00s)
✔ TemplatesTest: Duplicate (0.01s)
✔ TemplatesTest: Read for user (0.00s)
✔ TemplatesTest: Update (0.01s)
✔ TemplatesTest: Destroy (0.01s)
....
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 00:06.298, Memory: 68.00 MB

OK (149 tests, 192 assertions)
....
```

NicolasCARPi commented on src/controllers/ApiController.php:
```diff
@@ -613,6 +618,44 @@ private function createExperiment(): Response
+     */
+    private function createTemplate(): Response
+    {
+        if ($this->Entity instanceof Database) {
```
this check seems unneeded as there is already a check for instanceof Templates before calling this function

m6121 replied:
This was implemented analogue to the `createExperiment` function in the same controller. It might be cleaned there as well. Fixed in 4c4dbce

NicolasCARPi commented on src/classes/Update.php:
```diff
@@ -36,7 +36,7 @@
 {
     /** @var int REQUIRED_SCHEMA the current version of the database structure */
-    private const REQUIRED_SCHEMA = 57;
+    private const REQUIRED_SCHEMA = 58;
```
use schema57 and merge the code in schema57.sql

m6121 replied:
Changed in 4c4dbce

NicolasCARPi commented:
No idea why it's failing in circleci, but it's good that it doesn't fail locally at least.

Please address my requested changes, then I can test your code "for real" and merge this.

m6121 commented on src/maps/Team.php:
```diff
@@ -28,6 +28,13 @@ class Team implements MapInterface
 
     private int $id;
 
+    private string $commonTemplate = "<h1><span style='font-size: 14pt;'>Goal :</span></h1>
```
Changed in 4c4dbce

m6121 commented on src/models/Teams.php:
```diff
@@ -137,10 +137,19 @@ public function create(string $name): int
         $name = Filter::sanitize($name);
 
+        // default experiment template
+        $defaultBody = "<h1><span style='font-size: 14pt;'>Goal :</span></h1>
```
Implemented in 4c4dbce

m6121 commented on src/sql/schema58.sql:
```diff
@@ -0,0 +1,36 @@
+
+    ALTER TABLE `teams` ADD `common_template` text AFTER `name`;
+
+    UPDATE `teams` AS `m` SET `common_template` = (SELECT `body` FROM `experiments_templates` AS `a` WHERE `m`.`id` = `a`.`team` AND `a`.`title` = 'default' AND `a`.`userid` = 0);
```
Changed to full table names in 4c4dbce

m6121 commented on src/sql/schema58.sql:
```diff
@@ -0,0 +1,36 @@
+        `body` mediumtext NOT NULL,
+        `savedate` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+        `userid` int(10) UNSIGNED NOT NULL
+    ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
Implemented the table analogue to all other tables in `structure.sql`, all including the charset. Removed it in 4c4dbce

m6121 commented on src/sql/structure.sql:
```diff
@@ -220,6 +228,28 @@ CREATE TABLE `experiments_templates` (
+  `body` mediumtext NOT NULL,
+  `savedate` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `userid` int(10) UNSIGNED NOT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
Implemented the table analogue to all other tables in `structure.sql`, all including the charset. Removed it in 4c4dbce

m6121 commented on src/ts/ucp.ts:
```diff
@@ -20,10 +20,10 @@ $(document).ready(function() {
   $(document).on('click', '.createNewTemplate', function() {
-    const name = prompt('Template title');
-    if (name) {
+    const title = prompt('Template title');
```
Added the i18next implementation. However, the translations of course are not correct, but placeholders.

NicolasCARPi commented on src/models/Teams.php:
```diff
@@ -138,9 +138,10 @@ public function create(string $name): int
+        $sql = 'INSERT INTO teams (name, common_template, link_name, link_href) VALUES (:name, :common_template, :link_name, :link_href)';
         $req = $this->Db->prepare($sql);
         $req->bindParam(':name', $name);
+        $req->bindParam(':common_template', Templates::defaultBody);
```
I think it should be bindValue here.

See: Fatal error: Uncaught Error: PDOStatement::bindParam(): Argument #2 ($var) cannot be passed by reference in /home/scrutinizer/buil
/src/models/Teams.php on line 144

![2021-03-03-124138_953x414_scrot](https://user-images.githubusercontent.com/3043706/109800709-d555d500-7c1d-11eb-96ce-cf9e532baa3b.png)

m6121 replied:
Thanks for the suggested changed. Fixed in cfdd838

m6121 commented:
Just found one permission issue and noticed that the API get is not working as the show functions for `Templates` are different than in `AbstractEntity`. I think the implementation might probably be improved to make the permission checks inside the `read` function or align it a little more with `AbstractEntity`.

NicolasCARPi commented:
> I think the implementation might probably be improved to make the permission checks inside the read function or align it a little more with AbstractEntity.

Yes definitely! I think there are a lot of places in the code that can be factorized/better organized. I'll work on that in the next weeks, along with adding a new feature here and there. A lot of things were an afterthought. I would do things very differently if I had to start from scratch. But the good thing about code is that it can be remodeled *ad vitam aeternam*.

items_types should be items_templates, and should also depend from AbstractEntity, and the `Database` object should be `Items`, but hey, it's also the charm of historical baggage ;)

Now it is clear to me that it's easier to implement things for everything rather than doing things like "so this is only for experiments, and this is only for items". Makes the code much simpler, with less ifs when everyone has the same columns... That's why I added elabids to database items, and will surely add rating to experiments too.

NicolasCARPi reviewed (approved):
Ok it's good enough to be merged. I'll go over the changes and adjustments little by little. It's a better strategy to merge on a dev branch and get the new code in the codebase rather than trying to come up with the perfect PR ;)

Thanks for your work!

NicolasCARPi commented:
FYI, the new way of doing things for typescript actions: https://github.com/elabftw/elabftw/commit/3daeab94539f5269d234bc86058a7ef840241610

Still a WIP, as I'll probably merge all in Ajax.php, and Crud.class and Ajax.class will also be merged. See https://github.com/elabftw/elabftw/discussions/2513 ;)

NicolasCARPi commented:
Tests are working fine on scrutinizer-ci, and locally on both our machines, but fail on circleci and I have no idea why :man_shrugging: 
https://app.circleci.com/pipelines/github/elabftw/elabftw/2588/workflows/1254c41c-4131-4c3d-ad8b-8313ba6e2496/jobs/6424/parallel-runs/0/steps/0-122?invite=true

NicolasCARPi commented:
Spent some time moving from mariadb to mysql in circleci, but that's not the issue :/

Time to go to sleep anyway.

m6121 commented:
Thanks a lot for the fast and great feedback during the implementation of this feature! Will use the suggested improvements for potential future work.

Regarding Circle-CI: I was confused, becasue there was a step "Run unit tests" which passed for the last commits, however, the test afterwards called "Code coverage" as in your link failed: https://app.circleci.com/pipelines/github/elabftw/elabftw/2580/workflows/1402d229-0861-40f8-b7f5-1560791bbe0a/jobs/6416/parallel-runs/0/steps/0-121

NicolasCARPi commented:
Oh I think I get it now! The unit tests work, and it's when it's ran a second time to get coverage that it fails!!!! I didn't realize that before!

https://github.com/elabftw/elabftw/blob/hypernext/tests/run.sh#L43-L45

Yeah I had to do that because of a weird xdebug behavior. So now I can look into it!

NicolasCARPi commented:
Fixed the issue! It was necessary to "reset" the database to a known state before running the coverage again!

NicolasCARPi commented:
FYI here is the commit to add tests: https://github.com/elabftw/elabftw/commit/a8964a9b0010f7f481d6f1a131c9273e5d9bb305

m6121 commented:
That's great! Let me know if I can help anywhere.
